File: src/simulation/Event.java
package simulation;

import java.util.HashMap;
import java.util.Map;

/**
 * Object storing memory about something that happened in the world.
 */
public class Event {
    public enum Type{
        Death,
        ResourceDeath,
        AspectGaining,
        TileAcquisition,
        DisbandResources,
        GroupInteraction,

        Creation,
        Move,
        Change,

        Other
    }
    public Type type;
    public Map<String, Object> attributes = new HashMap<>();

    private String turn;
    public String description;

    public Event(Type type, String description, Object... attributes) {
        this(type, description);
        for (int i = 0; i < attributes.length; i += 2) {
            String name = (String) attributes[i];
            if (name.equals(""))
                break;
            this.attributes.put(name, attributes[i + 1]);
        }
    }

    public Event(Type type, String description, Map<String, Object> attributes) {
        this(type, description);
        this.attributes.putAll(attributes);
    }

    public Event(Type type, String description) {
        this.type = type;
        this.turn = Controller.session.world == null ? "None" : Controller.session.world.getTurn();
        this.description = description;
    }

    public Object getAttribute(String name) {
        return attributes.get(name);
    }

    @Override
    public String toString() {
        return turn + ". " + description;
    }
}

File: src/simulation/Event.kt
package simulation

import java.util.*


//Object storing memory about something that happened in the world.
class Event(var type: Type, val description: String) {
    enum class Type {
        Death,
        ResourceDeath,
        AspectGaining,
        TileAcquisition,
        DisbandResources,

        GroupInteraction,

        Creation,
        Move,
        Change,
        Other
    }

    private var _attributes: MutableMap<String, Any> = HashMap()
    val attributes: Map<String, Any>
        get() = _attributes

    private val turn: String = Controller.session.world?.getStringTurn() ?: "Pre-historic"

    constructor(type: Type, description: String, vararg attributes: Any) : this(type, description) {
        var i = 0
        while (i < attributes.size) {
            val name = attributes[i] as String
            if (name == "") break
            this._attributes[name] = attributes[i + 1]
            i += 2
        }
    }

    constructor(type: Type, description: String, attributes: Map<String, Any>) : this(type, description) {
        this._attributes.putAll(attributes)
    }

    fun getAttribute(name: String) = _attributes[name]

    override fun toString() = "$turn. $description"
}
